scripts/tank_cmd.py

```diff
@@ -1230,7 +1254,148 @@ def run_engine_cmd(log, pipeline_config_root, context_items, command, using_cwd,
+    if len(args) < 2:
+        raise InvalidCredentials("missing argument '%s'" % (arg1 if args[0][0] == arg2 else arg2))
+
+    # Make sure each arguments are not specified more than once.
+    _validate_only_once(args, arg1)
+    _validate_only_once(args, arg2)
+
+
+def _read_credentials_from_file(auth_path):
+    """
+    Reads the credentials from a file and returns two tuples: one for
+    login/password values and the other for script/key values. Any other values
+    are ignored.
+
+    :param auth_path: Path to a file with credentials.
+
+    :returns: A tuple of list of key, value pairs for values parsed. For example,
+       ([("login", "user"), ("password", "12345")],
+        [("script", "name"), ("key", "12345")])
+       The first tuple is concerned with human user based authentication, the second
+       is concerned with script based authentication.
+
+    :raises InvalidCredentials: If the file doesn't exist, this exception is raised.
+    """
+    if not os.path.exists(auth_path):
+        raise InvalidCredentials("credentials file does not exist.")
+    # Read the dictionary from file
+    with open(auth_path) as auth_file:
+        file_data = yaml.load(auth_file)
+
+    # We've already fixed the problem of parsing a bunch of values on the
+    # command line that may be mixed, so reshuffle the information that way
+    # and read from credentials
+    fake_cmd_line = ["--%s=%s" % (arg, value) for arg, value in file_data.iteritems()]
 
+    return (_extract_args(fake_cmd_line, ARG_SCRIPT, ARG_KEY),
+            _extract_args(fake_cmd_line, ARG_LOGIN, ARG_PASSWORD))
+
+
+def _extract_credentials(cmd_line):
+    """
+    Finds credentials from the command-line if the user specified them and
+    removes them from the array.
+
+    :param cmd_line: Array of arguments passed to the tank command.
+
+    :returns: A tuple of (filtered command line arguments, credentials).
+    """
+
+    # Extract the credential pairs.
+    script_user_credentials = _extract_args(cmd_line, ARG_SCRIPT, ARG_KEY)
+    human_user_credentials = _extract_args(cmd_line, ARG_LOGIN, ARG_PASSWORD)
+    file_credentials_path = _extract_args(cmd_line, ARG_AUTH)
+
+    # If we have credentials in a file
+    if file_credentials_path:
+        # make sure nothing else was specified on the command line.
+        if script_user_credentials or human_user_credentials:
+            raise InvalidCredentials("can't mix command line credentials and file credentials.")
+        # Validate it's only been specified once.
+        _validate_only_once(file_credentials_path, ARG_AUTH)
+        # Read the credentials from file
+        script_user_credentials, human_user_credentials = _read_credentials_from_file(
+            file_credentials_path[0][1]
+        )
+
+    # If we have credentials for both user types.
+    if script_user_credentials and human_user_credentials:
+        raise InvalidCredentials("can't mix human user and script user credentials.")
```

Comment:
I thought HumanUser was a generally accepted term from Shotgun, since we have ApiUser and HumanUser as entity types.